### ISSUE TITLE
DTLS handshake with "change cipher spec" out of order messes the TLS state machine

### ISSUE THREAD

**Issue by jpujol** [openssl/openssl#4929]
The following DTLS handshake sequence causes openSSL to generate a SSL_ERROR_SSL. In parentheses, the sequence numbers
```
peer1 -> peer2: ClientHello (0)
peer2 -> peer1: ServerHello (0), Certificate(1), ServerKeyExchange(2), ServerHelloDone(3)
peer1 -> peer2: ChangeCipherSpec(2)
peer1 -> peer2: ClientKeyExchange(1)
```
As soon as peer2 receives ChangeCipherSpec, it messes the state machine, instead of waiting for the message with sequence number 1

Any other messages out of order are processed correctly, thanks to the code in `statem_dtls.c`, lines 783:
```
    if (msg_hdr.seq != s->d1->handshake_read_seq) {
        *len = dtls1_process_out_of_seq_message(s, &msg_hdr, &ok);
        return ok;
    }
```
but before that, there is a check for ChangeCipherSpec, in which the sequence number is not checked

**Comment by mattcaswell** [comment]
> before that, there is a check for ChangeCipherSpec, in which the sequence number is not checked

Unfortunately there *is* no sequence number in a CCS message :-( (don't confuse the record sequence number and the message sequence number). I thought out-of-order CCS messages were just dropped...will need to investigate.

**Comment by jpujol** [comment]
[failing_case.pcapng.tar.gz](https://github.com/openssl/openssl/files/1558773/failing_case.pcapng.tar.gz)
Thanks for the reply. You are right, I was confusing both seq numbers... Will investigate. 
In case is of any help, please find attached the pcap of the failing case. Let me know if you need more information

**Comment by davidben** [comment]
BTW, I believe this is the scenario tested by the StrayChangeCipherSpec test in BoringSSL's test suite, though you all seem to have [disabled it](https://github.com/openssl/openssl/blob/master/test/ossl_shim/ossl_config.json#L189).

**Comment by mattcaswell** [comment]
> BTW, I believe this is the scenario tested by the StrayChangeCipherSpec test in BoringSSL's test suite, though you all seem to have disabled it.

Probably because it failed :-)

**Comment by jpujol** [comment]
yeah, just tried boringssl and it handles the situation properly

**Comment by mattcaswell** [comment]
This got auto-closed because I merged the master fix. Reopening until the 1.1.0 fix (in #6196) Also gets merged.

**Comment by mattcaswell** [comment]
#6196 got merged so closing.
